### DISCUSSION TITLE
milvus在search的时候进行混合搜索

### DISCUSSION THREAD

**Discussion by 131452099900** [milvus-io/milvus#30987] [Q&A and General discussion] [answered]
如果我使用的IVF_FLAT索引的，在我的searchr的时候nprobe没有全覆盖nlist有可能会漏过滤条件的数据吗？

**Comment by yhmo** [comment]
非常有可能。ANN search是找最相似的topk个结果，过滤条件是无脑地对全量数据过滤，符合过滤条件的不一定相似，而且有可能很不相似。

**Reply by 131452099900** [reply]
它的工作流程是不是先进行临近检索出topk然后在对这topk数据进行一个expr条件的过滤。如果我的nlist和nprobe一样的话其实是和query过滤一样的吧

**Reply by yhmo** [reply, marked as answer]
先用expr对全量数据进行过滤，然后再做邻近检索得出topk。如果nlist==nprobe，那么就跟使用FLAT做暴搜效果一样。
search()是做邻近检索。query()是取出符合过滤条件的数据。两者目的不一样，就算nlist==nprobe也不一样。假设一亿条数据里有1万条数据符合某个expr条件，search时找到topk=100，那就只返回100条最相似的，query是把1万条全部返回来。就算topk=10000，它们返回的结果也不一样，search返回的结果根据distance做了排序，query返回的结果是随机的。

**Reply by 131452099900** [reply]
好的了解了，谢谢

**Reply by 131452099900** [reply]
对了，我目前在我的单机milvus下面测试，采用的就是IVF_FLAT这种索引，数据量是10w，我现在在带expr的search（过滤掉的数据大概有9成）和不带expr的search，发现就是带expr可能更快一点，这种情况又怎么解释呢，之前好像说带expr会先进行全量数据的过滤。

**Reply by yhmo** [reply]
expr是全量过滤。
IVF是聚类索引，nlist参数决定了聚合成多少个簇，nprobe决定了搜索多少个簇。

假设10万条数据依次标记为 1, 2, 3, 4, ..... 99999, 100000


我的IVF_FLAT nlist=200，那么就有200个簇，每个簇有500条向量
我搜索时设置nprobe=10，那么就要计算10个簇里的向量

假设我设置一个expr，过滤掉 1, 2, 3, 4, ...... 90000，并且，我设置nlist，计算的10个簇里的向量都是1～90000之间的。那么，由于这10个簇里的向量都被过滤掉了，所以，一个向量都不需要计算，耗时为0。

如果我去掉expr，它就要计算这10个簇里的向量。所以，比设置expr要慢。

运行时搜索哪个簇是无法预期的，设置expr有可能比不设置expr快，也可能慢。
